File: src/main/java/com/bernd/game/PointSet.java
package com.bernd.game;

import static com.bernd.util.Util.divUp;

final class PointSet {

  private final int dim;
  private final int[] points;

  private PointSet(int dim, int[] points) {
    this.dim = dim;
    this.points = points;
  }

  static PointSet create(int dim) {
    return new PointSet(dim, new int[divUp(dim * dim, 0x20)]);
  }

  void add(int x, int y) {
    int ptId = y * dim + x;
    int pos = ptId >> 5;
    int test = 1 << (ptId & 0x1f);
    points[pos] = points[pos] | test;
  }

  boolean has(int x, int y) {
    int ptId = y * dim + x;
    int pos = ptId >> 5;
    int test = 1 << (ptId & 0x1f);
    return (points[pos] & test) != 0;
  }

  // visible for testing
  int length() {
    return points.length;
  }
}

File: src/main/client/src/model/PointSet.js
export class PointSet {

  constructor(dim) {
   this.dim = dim
   this.points = new Int32Array(Math.ceil((dim * dim) / 0x20)) 
  }

  add(x, y) {
    let ptId = y * this.dim + x
    let pos = ptId >> 5
    let test = 1 << (ptId & 0x1f)
    this.points[pos] |= test
  }

  has(x, y) {
    let ptId = y * this.dim + x
    let pos = ptId >> 5
    let test = 1 << (ptId & 0x1f)
    return (this.points[pos] & test) !== 0
  }
}
